### PR TITLE
Deprecate hash key path mapping

### PR DESCRIPTION
**Pull request by gmcgibbon**
Followup to https://github.com/rails/rails/pull/52409

### Motivation / Background

This Pull Request has been created because I'd like to deprecate paths as hash keys in route mappings:

```ruby
# Before
get "/users" => "users#index"
post "/logout" => :sessions

# After
get "/users", to: "users#index"
post "/logout", to: "sessions#logout"
```

This work moves us closer to [using keywords in the routing mapper](https://github.com/Shopify/rails/tree/routing_mapper_opt) which gives us a 1.25-1.5x speed boost (see linked PR above for benchmark).

### Detail

Drawing routes with has keys complicates route drawing enough to warrant deprecation. Transitioning the routing mapper to use keywords would result in a 1.25-1.5x improvement in speed, and it would be substantially easier to do if we drop this feature.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by gmcgibbon** [comment]
Looks like there's still some usages in docs and gems other than action pack I need to update.

File: actionpack/lib/action_dispatch/routing/mapper.rb [hunk 1]
```diff
@@ -637,6 +637,12 @@ def mount(app, options = nil)
             options = app
             app, path = options.find { |k, _| k.respond_to?(:call) }
             options.delete(app) if app
+
+            raise ArgumentError, "A rack application must be specified" unless app.respond_to?(:call)
```
Comment (p8) [inline actionpack/lib/action_dispatch/routing/mapper.rb:641]: This seems like an unrelated change?

File: actionpack/lib/action_dispatch/routing/mapper.rb [hunk 2]
```diff
@@ -637,6 +637,12 @@ def mount(app, options = nil)
...
+            ActionDispatch.deprecator.warn(<<-MSG.squish)
+              Mapping an engine route with a hash key name is deprecated and
+              will be removed in Rails 8.1. Please use the at: option instead.
```
Comment (p8) [inline actionpack/lib/action_dispatch/routing/mapper.rb:644]: Should this be `to:`?
```suggestion
              will be removed in Rails 8.1. Please use the to: option instead.
```

Comment (p8) [reply actionpack/lib/action_dispatch/routing/mapper.rb:644]: Ah nevermind. This is for mounting.

File: actionpack/lib/action_dispatch/routing/mapper.rb [hunk 3]
```diff
@@ -637,6 +637,12 @@ def mount(app, options = nil)
             options = app
             app, path = options.find { |k, _| k.respond_to?(:call) }
             options.delete(app) if app
+
+            raise ArgumentError, "A rack application must be specified" unless app.respond_to?(:call)
+            ActionDispatch.deprecator.warn(<<-MSG.squish)
+              Mapping an engine route with a hash key name is deprecated and
```
Comment (p8) [inline actionpack/lib/action_dispatch/routing/mapper.rb:643]: Maybe?
```suggestion
              Mounting an engine route with a hash key name is deprecated and
```

File: actionpack/lib/action_dispatch/routing/mapper.rb (hunk 1 above)
Comment (gmcgibbon) [reply actionpack/lib/action_dispatch/routing/mapper.rb:641]: I wanted to make the exception raise without a deprecation warning if you pass a hash without an app. I can try to think of another way that's more obvious, but maybe that's actually ok.

File: actionpack/lib/action_dispatch/routing/mapper.rb (hunk 3 above)
Comment (gmcgibbon) [reply actionpack/lib/action_dispatch/routing/mapper.rb:643]: Yeah, I think that's easier to understand language. I'll update to "Mounting" and "Drawing"

File: actionpack/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,17 @@
...
+    ```ruby
+    # Before
+    get "/users" => "users#index"
+    post "/logout" => :sessions
+
+    # After
+    get "/users", to: "users#index"
+    post "/logout", to: "sessions#logout"
+    ```
+
```
[the remark this answers is not in the document]
Comment (gmcgibbon) [reply actionpack/CHANGELOG.md:12]: Yes, will do! Thank you

**Comment by gmcgibbon** [comment]
There's deprecation warnings from sprockets, propshaft, and other engines from gems, but I can't fix them here. This should be good for now.

**Comment by chaadow** [comment]
If anybody has a custom cop (rubocop) I'd gladly appreciate it 💀 I have so many warnings in my production app' big `routes.rb` file

**Comment by tagliala** [comment]
For reference: PR reverted in 5c0b7496ab32c25c80f6d1bdc8b32ec6f75ce1e4
